### PR TITLE
test: report summary of failed tests

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Report formulae that fail their tests.
Example:
```
$ brew test adns gcc jmhobbs/parrot/terminal-parrot
Testing adns
==> /usr/local/Cellar/adns/1.5.1/bin/adnsheloex --version
Error: Testing requires the latest version of gcc
Testing jmhobbs/parrot/terminal-parrot
==> /usr/local/Cellar/terminal-parrot/1.1.0/bin/parrot -loops 1
Last 15 lines ...
[snip]
... 
Error: The following formulae failed to successfully complete the tests:
Error: jmhobbs/parrot/terminal-parrot
```
That is, only report formulae that are installed, linked and fail their tests.

### CODE REVIEW

File: Library/Homebrew/dev-cmd/test.rb [hunk 1]
```diff
@@ -86,7 +86,7 @@ def test
         ofail "#{f.full_name}: failed"
         puts e.message
         failed_formulae << f
-      rescue Exception => e
+      rescue StandardError => e
```
Comment (maxim-belkin) [inline Library/Homebrew/dev-cmd/test.rb]: I added this change to fix the following stylistic error:

```
$ brew style ./Library/Homebrew/dev-cmd/test.rb
== Library/Homebrew/dev-cmd/test.rb ==
W: 87:  7: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?

1 file inspected, 1 offense detected
```

**Comment by MikeMcQuaid** [comment]
Are there other command that do similarly? It occurs to me that this is inconsistent with how e.g. `install` might behave?

**Comment by maxim-belkin** [comment]
If a user issues a call to test a large number of formulae, these tests might take a while. With this PR, a user will a summary of which formulae failed tests. If `brew test` is executed in a CI, it will be easier to report which formulae failed.

`install` is different because when we're dealing with installations there might be a waterfall effect.

**Comment by MikeMcQuaid** [comment]
@maxim-belkin I'd like to see this implemented for e.g. `install`'s `caveats` before we do this for other commands.

File: Library/Homebrew/dev-cmd/test.rb [hunk 2]
```diff
@@ -84,12 +85,19 @@ def test
       rescue ::Test::Unit::AssertionFailedError => e
         ofail "#{f.full_name}: failed"
         puts e.message
-      rescue Exception => e # rubocop:disable Lint/RescueException
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/test.rb]: You're changing behaviour here, please change it back, thanks.

File: Library/Homebrew/dev-cmd/test.rb [hunk 3]
```diff
@@ -84,12 +85,19 @@ def test
...
         ofail "#{f.full_name}: failed"
         puts e, e.backtrace
+        failed_formulae << f
       ensure
         ENV.replace(env)
       end
     end
+    return false if failed_formulae.empty?
+    ofail <<-EOF.undent
+      The following formulae failed to complete the tests successfully:
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/test.rb]: This should not be output if a single formula was passed (which is the typical case which again makes me wonder if this is a useful change, I'm afraid).

**Pull request closed without merging**